### PR TITLE
fixes and improvements related to Core Data usage #1443

### PR DESCRIPTION
**Pull request by bryanmontz**
## Issues covered
#1443 

## Description
These changes are the result of the Core Data usage research prescribed by #1443. They are the changes I was confident making without further discussion. Other suggestions will be mentioned in the ticket.

1. This code: `Event().createIfNecessary(jsonEvent...` is creating Events and throwing them away immediately and causing the console error message "Failed to call designated initializer on NSManagedObject". The resolution is to make the `createIfNecessary...` function static.

2. Adding the launch argument `-com.apple.CoreData.ConcurrencyDebug 1` reveals that `Event.loadFirstQuotedNote()` is creating an event on the wrong thread. The resolution is to move the `Event.findOrCreateStubBy...` call to inside the `context.perform` call below. Reference https://useyourloaf.com/blog/debugging-core-data/.

3. In `RelayService.retryFailedPublishes()`, a JSONEvent was being created inside a nested loop. This caused needless extra work because it could be created only in the outer loop.

## How to test
These changes aren't user testable, but they are dev testable.
Running `main` (prior to the changes in this PR), you will see a message in the console like this:
"Failed to call designated initializer on NSManagedObject"

If you add the launch argument `-com.apple.CoreData.ConcurrencyDebug 1` to the scheme and run the code, you will get a crash (an assert really) if you scroll around in a feed view that contains quoted notes.

After the changes in this PR, both of those issues are resolved.

### CODE REVIEW

**Review by mplorentz — APPROVED** [review]
These changes all look good! I did a smoke test and didn't see any problems.

File: Nos/Service/Relay/RelayService.swift
Comment (martindsq) [inline Nos/Service/Relay/RelayService.swift:741]: I have two different feelings regarding this. On the one hand, the `for` looks easier to read, on the other hand, it is less computationally efficient. We are lopping through missedRelays and then looping through missedAddress.

What about a middle ground? That would mean 1) keep the movement  of `guard let jsonEvent = event.codable else { continue }` to the top as it is, but 2) keep doing `for missedRelay in missedRelays { guard let missedAddress = missedRelay.address else { continue } ... }` (without the `jsonEvent` clause).

File: Nos/Service/Relay/RelayService.swift
Comment (martindsq) [inline Nos/Service/Relay/RelayService.swift:387]: I'm fine from switching to self.persistenceController.parseContext  but I prefer it to be in just one line. For some reason, I never got used to multiline `if` clauses. If the line limit is an issue, solutions could be storing the result of .parse() in a variable just one line above would do or storing self.persistenceController.parseContext in a variable.

**Review by martindsq — CHANGES_REQUESTED** [review]
Awesome! It seems you caught a ton of potential issues here! I'm ok with all of these changes. I just left two minor comments. The PR looks fine even if you don't address them but I'd like to see your opinion.

File: Nos/Controller/PersistenceController.swift
Comment (joshuatbrown) [inline Nos/Controller/PersistenceController.swift:105]: I'd love it if we moved the DEBUG code to an extension. I'm totally fine with it living inside this file -- I just think it would be a little nicer in an extension. We have similar extensions for `SettingsView` and `DefaultFeatureFlags`.

File: Nos/Controller/PersistenceController.swift
Comment (joshuatbrown) [inline Nos/Controller/PersistenceController.swift:173]: Yes yes yes never stop fixing these! Thank you so much!

**Review by joshuatbrown — APPROVED** [review]
I love this! I agree with Martin about multiline `if` clauses and would prefer to see that changed, but these are big improvements and this is 100% shippable from my perspective.

File: Nos/Service/Relay/RelayService.swift
Comment (bryanmontz) [reply Nos/Service/Relay/RelayService.swift:741]: This is all just my opinion, so I'd like to leave it as is, but...
Since the array of missedRelays is always going to be relatively small (how many relays can a user have at maximum?), it is a worthwhile tradeoff to make for readability. And generally I consider a `guard` at the top of the body of a loop to be a code smell because there are other mechanisms for filtering a collection, such as a `where` clause in your for loop (`for thing in things where thing.someCondition`). Optimizing for performance here is premature in my opinion.

File: Nos/Controller/PersistenceController.swift
Comment (bryanmontz) [reply Nos/Controller/PersistenceController.swift:105]: Yeah, that's nicer. Thanks!

**Comment by mplorentz** [comment]
@martindsq I think in this case posting your review as a Comment rather than Request Changes would have sped things along. Right now merging is blocked waiting on your approval even though you said you were fine with all changes. I know we have had some debate about when to use Request Changes vs Comment so I wanted to point this out.

**Review by martindsq — APPROVED** [review]

**Pull request merged**
